request/request.go

```diff
@@ -0,0 +1,148 @@
+package request
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+
+	"github.com/amp-labs/cli/logger"
+)
+
+type RequestClient struct {
+	Client *http.Client
+}
+
+// Header is a key/value pair that can be added to a request.
+type Header struct {
+	Key   string
+	Value string
+}
+
+func NewRequestClient() *RequestClient {
+	return &RequestClient{Client: http.DefaultClient}
+}
+
+// Put makes a PUT request to the desired URL, and unmarshalls the
+// response body into `result`.
+func (c *RequestClient) Put(ctx context.Context,
+	url string, reqBody any, result any, headers ...Header,
+) (*http.Response, error) {
+	req, err := makeJSONPutRequest(ctx, url, headers, reqBody)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.makeRequestAndParseResult(req, result)
+}
+
+// Post makes a POST request to the desired URL, and unmarshalls the
+// response body into `result`.
+func (c *RequestClient) Post(ctx context.Context,
+	url string, reqBody any, result any, headers ...Header,
+) (*http.Response, error) {
+	req, err := makeJSONPostRequest(ctx, url, headers, reqBody)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.makeRequestAndParseResult(req, result)
+}
+
+var ErrNone200Status = errors.New("error response from API")
+
+func (c *RequestClient) makeRequestAndParseResult(req *http.Request, result any) (*http.Response, error) {
+	logger.Debugf(">>> API REQUEST:\n%+v\n", req)
+
+	res, payload, err := c.sendRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Debugf("<<< API RESPONSE:\n%+v\n", res)
+
+	if res.StatusCode < 200 || res.StatusCode > 299 {
+		return res, fmt.Errorf("%w: HTTP Status %s", ErrNone200Status, res.Status)
+	}
+
+	if err := json.Unmarshal(payload, result); err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func makeJSONPostRequest(ctx context.Context, url string, headers []Header, body any) (*http.Request, error) {
+	jBody, err := json.Marshal(body)
+	if err != nil {
+		return nil, fmt.Errorf("request body is not valid JSON, body is %v:\n%w", body, err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewBuffer(jBody))
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+
+	headers = append(headers, Header{Key: "Content-Type", Value: "application/json"})
+	req.ContentLength = int64(len(jBody))
+
+	return addAcceptJSONHeaders(req, headers)
+}
+
+func makeJSONPutRequest(ctx context.Context, url string, headers []Header, body any) (*http.Request, error) {
+	jBody, err := json.Marshal(body)
+	if err != nil {
+		return nil, fmt.Errorf("request body is not valid JSON, body is %v:\n%w", body, err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, url, bytes.NewBuffer(jBody))
```

Comment:
I'd use `bytes.NewReader` instead of `bytes.NewBuffer`. It might be fine this way, but my thinking is that reader == "read only" whereas buffer == "read+write". Minor semantic difference, I know, and probably too pedantic to actually matter. But it makes the intention more clear, at least in my mind. Up to you, I'm not gonna die on this hill :P

Reply:
good call.